src/pyhf/events.py

```diff
@@ -6,10 +8,18 @@ def noop(*args, **kwargs):
     pass
 
 
-class Callables(list):
+class WeakList(list):
+    def append(self, item):
+        list.append(self, weakref.WeakMethod(item, self.remove))
+
+
+class Callables(WeakList):
     def __call__(self, *args, **kwargs):
-        for f in self:
-            f(*args, **kwargs)
+        for func in self:
+            if isinstance(func, weakref.WeakMethod):
+                func()(*args, **kwargs)
+            else:
+                func(*args, **kwargs)
```

Comment:
We're seeing a drop in coverage here: https://codecov.io/gh/scikit-hep/pyhf/pull/1035/diff?src=pr&el=tree#diff-c3JjL3B5aGYvZXZlbnRzLnB5

Would be good to get this covered.